lighthouse-core/gather/driver.js

```diff
@@ -74,6 +73,10 @@ class Driver {
       if (this._networkStatusMonitor) {
         this._networkStatusMonitor.dispatch(event);
       }
+
+      // @ts-ignore TODO(bckenny): tsc can't type event.params correctly yet,
+      // typing as property of union instead of narrowing from union of
+      // properties. See https://github.com/Microsoft/TypeScript/pull/22348.
```

Comment:
this is unfortunate to ignore, but will have to wait on ts team for the fix.

Basically `_eventEmitter.emit()` correctly checks that for every protocol event `method` name (e.g. `'Profiler.consoleProfileStarted'`), it will have a matching event `params` (e.g. `LH.Crdp.Profiler.ConsoleProfileStartedEvent`).

The `event` object passed into this function is a giant union of possible `method`/`pair` objects like:
```js
{method: 'Profiler.consoleProfileStarted', params: LH.Crdp.Profiler.ConsoleProfileStartedEvent} |
{method: 'Runtime.bindingCalled', params: LH.Crdp.Runtime.BindingCalledEvent} |
{method: 'CSS.fontsUpdated', params: LH.Crdp.CSS.FontsUpdatedEvent} |
...
```

It's a [discriminated union](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions), so in theory the compiler could check each union member separately, which would pass the `_eventEmitter.emit()` check.

Currently, however, the compiler doesn't narrow the type based on `method`, so instead of taking each union member separately, it takes the union of all possible property values. The second argument passed to `_eventEmitter.emit()` ends up being `LH.Crdp.Profiler.ConsoleProfileStartedEvent | LH.Crdp.Runtime.BindingCalledEvent | LH.Crdp.CSS.FontsUpdatedEvent | ...`.

When `_eventEmitter.emit()` starts checking that for every `method` (`'Profiler.consoleProfileStarted'`) there is the correct `params` (`LH.Crdp.Profiler.ConsoleProfileStartedEvent`), it sees that full union instead and so rightly shows an error since only one member of the union would have been the correct type.

As noted, this will hopefully be fixed by https://github.com/Microsoft/TypeScript/pull/22348, currently aimed at the typescript 3.2ish timeframe.

Reply:
Also, it's slightly confusing, but the similar checks in the surrounding methods (e.g. other uses of `eventEmitter`) *are* correctly type checked. The key is that `event` is an object with `method` and `params` as properties. If `method` and `params` had come in as separate function parameters (as they do elsewhere) or as a tuple, the compiler would be able to correctly type them.